jumanji/environments/logic/sudoku/generator.py

```diff
@@ -0,0 +1,145 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.i
+
+import abc
+import os
+from typing import Optional
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.logic.sudoku.constants import BOARD_WIDTH
+from jumanji.environments.logic.sudoku.types import State
+from jumanji.environments.logic.sudoku.utils import update_action_mask
+
+DATABASES = {
+    "very-easy": "data/1000_very_easy_puzzles.npy",  # 1000 puzzles with >= 46 clues
+    "mixed": "data/10000_mixed_puzzles.npy",  # 10000 puzzles with a random number
+    # of clues
+}
+
+
+class Generator(abc.ABC):
+    @abc.abstractmethod
+    def __call__(self, key: chex.PRNGKey) -> State:
+        """Call method responsible for generating a new state.
+
+        Args:
+            key: jax random key for any stochasticity used in the instance
+                generation process.
+
+        Returns:
+            A Sudoku State.
+        """
+        pass
+
+
+class DummyGenerator(Generator):
+    """Generates always the same board, for debugging purpose."""
+
+    def __init__(
+        self,
+    ) -> None:
+        board = jnp.array(
+            [
+                [0, 0, 0, 8, 0, 1, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 4, 3],
+                [5, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 7, 0, 8, 0, 0],
+                [0, 0, 0, 0, 0, 0, 1, 0, 0],
+                [0, 2, 0, 0, 3, 0, 0, 0, 0],
+                [6, 0, 0, 0, 0, 0, 0, 7, 5],
+                [0, 0, 3, 4, 0, 0, 0, 0, 0],
+                [0, 0, 0, 2, 0, 0, 6, 0, 0],
+            ]
+        )
+        solved_board = jnp.array(
+            [
+                [2, 3, 7, 8, 4, 1, 5, 6, 9],
+                [1, 8, 6, 7, 9, 5, 2, 4, 3],
+                [5, 9, 4, 3, 2, 6, 7, 1, 8],
+                [3, 1, 5, 6, 7, 4, 8, 9, 2],
+                [4, 6, 9, 5, 8, 2, 1, 3, 7],
+                [7, 2, 8, 1, 3, 9, 4, 5, 6],
+                [6, 4, 2, 9, 1, 8, 3, 7, 5],
+                [8, 5, 3, 4, 6, 7, 9, 2, 1],
+                [9, 7, 1, 2, 5, 3, 6, 8, 4],
+            ]
+        )
+
+        action_mask = board != 0
+        board = jnp.array(board, dtype=jnp.int32) - 1
+        action_mask = jnp.array(action_mask, dtype=jnp.int32)
+        action_mask = 1 - jnp.expand_dims(action_mask, -1).repeat(BOARD_WIDTH, axis=-1)
+        action_mask = update_action_mask(action_mask, board).astype(bool)
+        self._solved_board = jnp.array(solved_board, dtype=jnp.int32) - 1
+        self._board = board
+        self._action_mask = action_mask
+
+    def __call__(self, key: chex.PRNGKey) -> State:
+        return State(board=self._board, action_mask=self._action_mask, key=key)
+
+
+class DatabaseGenerator(Generator):
+    """Generates a board by sampling uniformly inside a puzzle database"""
+
+    def __init__(
+        self,
+        level: str = "mixed",
+        custom_boards: Optional[jnp.ndarray] = None,
```

Comment:
Would you prefer someting like this ?
```python
    def __init__(
        self,
        dataset: Optional[chex.Array] = None 
    ):
      if dataset is None:
          file_path = os.path.dirname(os.path.abspath(__file__))
          database_file = DATABASES["mixed"]
          self._boards = jnp.load(os.path.join(file_path, database_file))
      else:
          self._boards = dataset
  
 ```

Reply:
Yes, this looks good to me with the only modification
```
    def __init__(
        self,
        dataset: Optional[chex.Array], 
    ):
```
Because we don't necessarily want to define default behaviors for the generator.